### PR TITLE
[10157] Always write disttrial log file in UTF8.

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

Right now the distrial test.log is written in default encoding.

The default encoding can be cp1252 on windows.

The sub-process sends logs data as unicode via AMP and with cp1252 encoding things break as the Unicode data sent by the sub-process can't be represented as CP1252/Latin1

This error was found while running `trial -j` on Windows and a test was generating log messages with unicode data.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/
* [NO] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: adiroiban
Reviewer: 
Fixes: ticket:10157

Always write distrial log file in UTF-8.
```

### CODE REVIEW

File: src/twisted/trial/_dist/test/test_worker.py [hunk 1]
```diff
@@ -331,19 +331,49 @@ def test_childDataReceived(self):
...
+        logFile = tempDir.child("test.log")
+
+        def getLatin1(do_setlocale: bool = True) -> str:
+            """
+            A replacement for L{locale.getpreferredencoding} that always
+            returns the Latin-1 encoding.
+            """
+            return "latin-1"
+
+        self.patch(_bootlocale, "getpreferredencoding", getLatin1)
```
Comment (adiroiban) [inline src/twisted/trial/_dist/test/test_worker.py]: I am happy to write this test in another way.
I have observed that patching just `locale` was not triggering the expected TDD initial test failure.

File: src/twisted/trial/_dist/test/test_worker.py [hunk 2]
```diff
@@ -272,9 +274,10 @@ def stopTest(self, test):
         return d.addCallback(self.assertIdentical, result)
 
 
-class FakeAMProtocol(AMP):
+class SpyDataLocalWorkerAMP(LocalWorkerAMP):
```
Comment (adiroiban) [inline src/twisted/trial/_dist/test/test_worker.py:277]: I went with using `LocalWorkerAMP` as much as possible to test the real deal and reduce code duplication in tests.

**Comment by adiroiban** [comment]
This is ready for an initial review.

File: src/twisted/newsfragments/10157.bugfix [hunk 3]
```diff
@@ -0,0 +1,6 @@
+twisted.trial._dist.worker.LocalWorker.connectionMade now always writes the
+log file using UTF-8 encoding.
+In previous versions it was using the system default encoding.
+This was causing encoding errors as the distributed trial workers are sending
+Unicode data and the system default encoding might now always be Unicode compatible.
```
Comment (wiml) [inline src/twisted/newsfragments/10157.bugfix]:
```suggestion
Unicode data and the system default encoding might not always be Unicode compatible.
```

**Review by wiml — APPROVED** [review]
LGTM. (Though, noticed a typo in the newsfragment.)

**Pull request merged**